module_3/gift_certificates_system/model/src/main/java/com/epam/esm/listener/GiftCertificateToTagRelationListener.java

```diff
@@ -0,0 +1,29 @@
+package com.epam.esm.listener;
+
+import com.epam.esm.entity.GiftCertificate;
+import com.epam.esm.entity.GiftCertificateToTagRelation;
+
+import javax.persistence.PrePersist;
+import javax.persistence.PreRemove;
+import javax.persistence.PreUpdate;
+import java.time.LocalDateTime;
+
+public final class GiftCertificateToTagRelationListener {
+    private static LocalDateTime currentDateTime;
+
+    @PrePersist
+    public void beforeCreate(GiftCertificateToTagRelation relation) {
+        setGiftCertificateLastUpdateTime(relation);
+    }
+
+    @PreRemove
+    public void beforeRemove(GiftCertificateToTagRelation relation) {
+        setGiftCertificateLastUpdateTime(relation);
+    }
+
+    private void setGiftCertificateLastUpdateTime(GiftCertificateToTagRelation relation) {
+        currentDateTime = LocalDateTime.now();
+        GiftCertificate giftCertificate = relation.getGiftCertificate();
+        giftCertificate.setLastUpdateDate(currentDateTime);
+    }
```

Comment:
Is it thread safe? I suppose, no

Reply:
Thanks, fixed ([see](https://github.com/n199a/epam_training_center/pull/3#discussion_r751191824))